# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/test/webgl-conf/checkout/conformance/context/context-lost.html
# Commit: 8e0aea33f622
# Full Hash: 8e0aea33f622a76c5c8baa7dcf4485ffefd1662f
# Author: Kelsey Gilbert <kelsey.gilbert@mozilla.com>
# Date: 2024-07-09 09:45:28
# Regressor Bug: 1892542
# File Overlap Count: 1
# Description:
#   Bug 1892542 - Webgl e.g. createBuffer() made infallible. r=gfx-reviewers,webidl,saschanaz,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D208050
# ==============================================================================

diff -r b953e0ccb68b -r 8e0aea33f622 dom/canvas/test/webgl-conf/checkout/conformance/context/context-lost.html
--- a/dom/canvas/test/webgl-conf/checkout/conformance/context/context-lost.html	Mon Jul 08 19:36:12 2024 +0300
+++ b/dom/canvas/test/webgl-conf/checkout/conformance/context/context-lost.html	Mon Jul 08 17:24:12 2024 +0000
@@ -308,12 +308,6 @@
 
     // Functions return nullable values should all return null.
     var nullTests = [
-        "gl.createBuffer()",
-        "gl.createFramebuffer()",
-        "gl.createProgram()",
-        "gl.createRenderbuffer()",
-        "gl.createShader(gl.GL_VERTEX_SHADER)",
-        "gl.createTexture()",
         "gl.getActiveAttrib(program, 0)",
         "gl.getActiveUniform(program, 0)",
         "gl.getAttachedShaders(program)",
@@ -336,14 +330,24 @@
     ];
     testFunctionsThatReturnNULL(nullTests);
 
+    [
+      [() => gl.createBuffer(), x => gl.bindBuffer(gl.ARRAY_BUFFER, x), x => gl.isBuffer(x)],
+      [() => gl.createFramebuffer(), x => gl.bindFramebuffer(gl.FRAMEBUFFER, x), x => gl.isFramebuffer(x)],
+      [() => gl.createProgram(), x => undefined, x => gl.isProgram(x)],
+      [() => gl.createRenderbuffer(), x => gl.bindRenderbuffer(gl.RENDERBUFFER, x), x => gl.isRenderbuffer(x)],
+      [() => gl.createShader(gl.VERTEX_SHADER), x => undefined, x => gl.isShader(x)],
+      [() => gl.createTexture(), x => gl.bindTexture(gl.TEXTURE_2D, x), x => gl.isTexture(x)],
+    ].forEach(([fn_create, fn_bind, fn_is]) => {
+      const x = fn_create();
+      assertMsg(x, `${fn_create.toString()} -> non-null`);
+      assertMsg(fn_is(x) === false, `[before bind] ${fn_is.toString()} -> false`);
+      fn_bind(x);
+      fn_bind(null);
+      assertMsg(fn_is(x) === false, `[after bind] ${fn_is.toString()} -> false`);
+    });
+
     // "Is" queries should all return false.
-    shouldBeFalse("gl.isBuffer(buffer)");
     shouldBeFalse("gl.isEnabled(gl.BLEND)");
-    shouldBeFalse("gl.isFramebuffer(framebuffer)");
-    shouldBeFalse("gl.isProgram(program)");
-    shouldBeFalse("gl.isRenderbuffer(renderbuffer)");
-    shouldBeFalse("gl.isShader(shader)");
-    shouldBeFalse("gl.isTexture(texture)");
 
     shouldBe("gl.getError()", "gl.NO_ERROR");
 
@@ -355,10 +359,17 @@
           "OES_vertex_array_object.isVertexArrayOES(vertexArrayObject)",
           "OES_vertex_array_object.deleteVertexArrayOES(vertexArrayObject)",
         ]);
-      testFunctionsThatReturnNULL(
-        [
-          "OES_vertex_array_object.createVertexArrayOES()",
-        ]);
+
+      [
+        [() => OES_vertex_array_object.createVertexArrayOES(), x => OES_vertex_array_object.isVertexArrayOES(x), x => OES_vertex_array_object.isVertexArrayOES(x)],
+      ].forEach(([fn_create, fn_bind, fn_is]) => {
+        const x = fn_create();
+        assertMsg(x, `${fn_create.toString()} -> non-null`);
+        assertMsg(fn_is(x) === false, `[before bind] ${fn_is.toString()} -> false`);
+        fn_bind(x);
+        fn_bind(null);
+        assertMsg(fn_is(x) === false, `[after bind] ${fn_is.toString()} -> false`);
+      });
     }
 
     testUploadingLostContextToTexture();